crates/full-node/sov-sequencer/src/preferred/tx_nonce_queue.rs

```diff
@@ -0,0 +1,722 @@
+    /// A nonce is considered "present" if it's either in the queue OR marked as last_executed.
+    fn has_contiguous_sequence_to(&self, tx_nonce: u64, current_nonce: u64) -> bool {
+        // We know the user account's nonce cannot be lower than the state value.
+        // Sometimes the state value is stale for a short period of time, which is why we treat is
+        // as a lower bound only.
+        let lower_bound = current_nonce;
+        // last_executed is the last recorded transaction popped from the user's queue, so we know
+        // the account's nonce cannot be higher than this.
+        // We add 1 because `current_nonce` is the next valid nonce, while `last_executed` was the
+        // previous valid nonce (so the next transaction should have nonce `last_executed + 1`).
+        let upper_bound =
+            current_nonce.max(self.last_executed.map(|n| n.saturating_add(1)).unwrap_or(0));
+
+        if tx_nonce < lower_bound {
+            // The transaction can never be valid.
+            false
+        } else if lower_bound <= tx_nonce && tx_nonce <= upper_bound {
+            // There is uncertainty about the user account's real nonce. In this range, leniently
+            // treat transactions as valid (the uncertainty will eventually resolve itself as API
+            // state updates).
+            true
+        } else if tx_nonce > upper_bound {
+            // The transaction's nonce is known to be greater than our upper bound estimate of the
+            // next valid nonce.
+            // We need to check if there's a contiguous set of transactions actually queued beyond
+            // the upper bound.
+            let mut expected = upper_bound;
+            for &nonce in self.txs.keys() {
+                if nonce >= tx_nonce {
+                    break;
+                }
+                if nonce != expected {
+                    return false; // Gap found
+                }
+                expected += 1;
+            }
+
+            // Return true if we've reached tx_nonce, meaning we have all prerequisites
+            expected >= tx_nonce
+        } else {
+            unreachable!("The ranges checked are meant to be exhaustive");
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.txs.is_empty()
+    }
+}
+
+#[derive(Default)]
+pub struct TxNonceQueues<S: Spec, Rt: Runtime<S>> {
+    queues: DashMap<CredentialId, AddressQueue<S, Rt>>,
+}
+
+#[allow(dead_code)]
+impl<S: Spec, Rt: Runtime<S>> TxNonceQueues<S, Rt> {
+    pub fn new() -> Self {
+        Self {
+            queues: DashMap::new(),
+        }
+    }
+
+    /// Lock the queue for a specific address for atomic nonce check + enqueue
+    pub fn lock_for_address(
+        &self,
+        credential_id: CredentialId,
+    ) -> dashmap::mapref::entry::Entry<CredentialId, AddressQueue<S, Rt>> {
+        self.queues.entry(credential_id)
```

Comment:
There’s something I don’t quite understand about DashMap. The documentation for `entry` says:

```
Locking behaviour: May deadlock if called when holding any sort of reference into the map.
```

Does this mean that if one thread has already locked an entry, and another thread tries to lock it again, it could lead to a deadlock?
My understanding is that the first thread would still be holding a reference into the map’s internal structure, which might be causing this issue. But maybe I don't read the documentation properly WDT?

Reply:
Yeah the documentation is confusing, I thought the same too at first. But it says at the top
```
Documentation mentioning locking behaviour acts in the reference frame of the calling thread. This means that it is safe to ignore it across multiple threads.
```
So it's safe for multithreading, you just have to be careful within a single thread